### PR TITLE
bash-completion: analyze: complete system call sets

### PR DESCRIPTION
**Pull request by lucaswerkmeister**
---

I didn’t add any zsh support because I’m not familiar enough with zsh – sorry.

### CODE REVIEW

File: shell-completion/bash/systemd-analyze [hunk 1]
```diff
@@ -30,6 +30,15 @@ __get_machines() {
         machinectl list --no-legend --no-pager | { while read a b; do echo " $a"; done; };
 }
 
+__get_syscall_sets() {
+        local line
+        systemd-analyze syscall-filter --no-pager | while IFS= read -r line; do
+                if [[ $line == @* ]]; then
+                        printf '%s\n' "$line"
```
Comment (yuwata) [inline shell-completion/bash/systemd-analyze:37]: I guess that `echo " $line"` is faster.

File: shell-completion/bash/systemd-analyze [hunk 2]
```diff
@@ -30,6 +30,15 @@ __get_machines() {
         machinectl list --no-legend --no-pager | { while read a b; do echo " $a"; done; };
 }
 
+__get_syscall_sets() {
+        local line
+        systemd-analyze syscall-filter --no-pager | while IFS= read -r line; do
```
Comment (poettering) [inline shell-completion/bash/systemd-analyze:35]: actually, I am pretty sure the `--no-pager` thing is just cargo-culting here... it's the implied default if our tools detect they are run as part of a shell pipeline (i.e. stdout is not a tty)... hence it should be safe to remove it here and everywhere else where we make it part of a shell pipeline..

**Comment by poettering** [comment]
looks superficially OK to me, but I'll leave it to others (@keszybz ?) who are more at home in shell completion logic to review this finally. Also @yuwata's comment has not been addressed yet...

File: shell-completion/bash/systemd-analyze (hunk 1 above)
Comment (lucaswerkmeister) [reply shell-completion/bash/systemd-analyze:37]: I couldn’t measure any significant difference:

```sh
$ for i in {1..1000000}; do printf '%s\n' "@foo" " bar" " baz"; done > /tmp/output
$ time while IFS= read -r line; do if [[ $line == @* ]]; then printf '%s\n' "$line"; fi; done < /tmp/output > /dev/null

real    1m18,929s
user    1m13,667s
sys     0m5,232s
$ time while IFS= read -r line; do if [[ $line == @* ]]; then echo " $line"; fi; done < /tmp/output > /dev/null

real    1m14,851s
user    1m10,103s
sys     0m4,741s
$ time while IFS= read -r line; do if [[ $line == @* ]]; then printf '%s\n' "$line"; fi; done < /tmp/output > /dev/null

real    1m15,528s
user    1m10,841s
sys     0m4,665s
$ time while IFS= read -r line; do if [[ $line == @* ]]; then echo " $line"; fi; done < /tmp/output > /dev/null

real    1m33,447s
user    1m27,709s
sys     0m5,655s
$ time while IFS= read -r line; do if [[ $line == @* ]]; then printf '%s\n' "$line"; fi; done < /tmp/output > /dev/null

real    1m45,044s
user    1m38,670s
sys     0m6,281s
$ time while IFS= read -r line; do if [[ $line == @* ]]; then echo " $line"; fi; done < /tmp/output > /dev/null

real    1m18,719s
user    1m13,438s
sys     0m5,259s
```

I think the differences here are just due to other load on the system where I was running this.

My reason for preferring `printf` over `echo` is that `echo` is more fragile in general. It’s not as much of a problem here – it’s a bash completion script, and in bash `echo` slightly more sane than in some other shells – but in general, I consider `echo` to be a minor code smell in shell scripts, and use `printf` whenever writing them myself.

Comment (keszybz) [reply shell-completion/bash/systemd-analyze:37]: Yeah, I think `printf` is considered more portable in general.

File: shell-completion/bash/systemd-analyze (hunk 2 above)
Comment (keszybz) [reply shell-completion/bash/systemd-analyze:35]: Yeah. We can probably do that in one fell swoop though. Doesn't matter to much here.

**Pull request merged**